griptape/engines/eval/eval_engine.py

```diff
@@ -0,0 +1,137 @@
+        if value is None:
+            if self.criteria is None:
+                raise ValueError("either evaluation_steps or criteria must be specified")
+            return
+
+        if self.criteria is not None:
+            raise ValueError("can't have both evaluation_steps and criteria specified")
+
+        if not value:
+            raise ValueError("evaluation_steps must not be empty")
+
+    def evaluate(self, input: str, actual_output: str, **kwargs) -> tuple[float, str]:  # noqa: A002
+        evaluation_params = {
+            key.replace("_", " ").title(): value
+            for key, value in {"input": input, "actual_output": actual_output, **kwargs}.items()
+        }
+
+        if self.evaluation_steps is None:
+            with validators.disabled():
+                self.evaluation_steps = self._generate_steps(evaluation_params)
+
+        return self._generate_results(evaluation_params)
+
+    def _generate_steps(self, evaluation_params: dict[str, str]) -> list[str]:
+        system_prompt = self.generate_steps_system_template.render(
+            evaluation_params=self.__generate_evaluation_params_text(evaluation_params),
+            criteria=self.criteria,
+            json_schema_rule=JsonSchemaRule(STEPS_SCHEMA.json_schema("Output Format")),
+        )
+        user_prompt = self.generate_steps_user_template.render()
+
+        result = self.prompt_driver.run(
+            PromptStack(
+                messages=[
+                    Message(system_prompt, role=Message.SYSTEM_ROLE),
+                    Message(user_prompt, role=Message.USER_ROLE),
+                ],
+            ),
+        ).to_artifact()
+
+        parsed_result = json.loads(result.value)
+
+        return parsed_result["steps"]
+
+    def _generate_results(self, evaluation_params: dict[str, str]) -> tuple[float, str]:
+        system_prompt = self.generate_results_system_template.render(
+            evaluation_params=self.__generate_evaluation_params_text(evaluation_params),
+            evaluation_steps=self.evaluation_steps,
+            evaluation_text=self.__generate_evaluation_text(evaluation_params),
+            json_schema_rule=JsonSchemaRule(RESULTS_SCHEMA.json_schema("Output Format")),
+        )
+        user_prompt = self.generate_results_user_template.render()
+
+        result = self.prompt_driver.run(
+            PromptStack(
+                messages=[
+                    Message(system_prompt, role=Message.SYSTEM_ROLE),
+                    Message(user_prompt, role=Message.USER_ROLE),
+                ],
+            ),
+        ).to_artifact()
+
+        parsed_result = json.loads(result.value)
+
+        score = float(parsed_result["score"]) / 10
+        reason = parsed_result["reason"]
+
+        return score, reason
+
+    def __generate_evaluation_params_text(self, evaluation_params: dict[str, str]) -> str:
+        return ", ".join(param for param in evaluation_params)
```

Comment:
Same as `", ".join(evalutation_params.keys())`? Almost feel like it'd be more readable to inline even if adds duplication since you really need to know its implementation to know whether or not you can reuse it (and it's really short anyway).